Add interpolation parameter to resize_to_po2()

theoratkin opened the pull request:
`Image::resize_to_po2()` now takes an optional `p_interpolation` parameter that it passes directly to `resize()` with default value `INTERPOLATE_BILINEAR`.

Can be used in future to prevent #44379 in 4.0.

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

theogen-ratkin commented:
Whoops, forgot to add a closing `</argument>` tag in the documentation.

clayjohn commented:
I just realized this is compatibility breaking. For 4.0 it makes sense and is IMO a better default. For 3.2 you'll have to make another PR that uses LINEAR as the default but uses NEAREST when called from rasterizer_storage.

lawnjelly commented:
Will using nearest filter as a default for power of 2 also affect the textures in 3d games (which may not want a pixellated look?).

I'm wondering whether we could use the `TEXTURE_FLAG_FILTER` flag to determine whether to use nearest filtering or not, giving the most appropriate result in both types of games.

theogen-ratkin commented:
@lawnjelly But textures with filter flag get filtered regardless, it's only pixelated at resize time. Does it make sense to filter textures twice?

@clayjohn No problem, I'll send another PR.

lawnjelly commented:
> @lawnjelly But textures with filter flag get filtered regardless, it's only pixelated at resize time. Does it make sense to filter textures twice?

Filtering at each stage is fairly standard for textures that aren't intended to be pixellated. Even mipmapping chain generation typically does e.g. a box filter.

Filtering by the API in realtime is texture size dependent, and cannot always counteract pixellation introduced in the source texture (and is not designed to).

theogen-ratkin commented:
Okay, it makes sense. Should we do that in a separate PR, since it's technically another change based on this PR?

theogen-ratkin commented:
@clayjohn 
I created another version of this PR that doesn't break 3.2 compatibility. #44460

YeldhamDev commented:
Those commits should be squash. See the instructions [in the docs](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html?highlight=git%20rebase#the-interactive-rebase).

theogen-ratkin commented:
@YeldhamDev Done!

clayjohn commented:
Lawnjelly and I discussed this earlier today and came to the conclusion that the default should actually be linear. While it is true that in your example case it needs to be nearest, non-pixel-art textures will have significant artifacts introduced if using nearest.

lawnjelly commented:
> Lawnjelly and I discussed this earlier today and came to the conclusion that the default should actually be linear. While it is true that in your example case it needs to be nearest, non-pixel-art textures will have significant artifacts introduced if using nearest.

Yes just to clarify - the approach should be fine in your other 3.2 PR provided you respect the filter flag (that is set in texture import). But in master, GLES2 isn't even fully implemented yet, the other APIs support non POT, and the mechanisms for filtering / wrapping are changing, so it would be premature to change there.

theogen-ratkin commented:
Okay, maybe then just set filtering to bilinear by default, since this optional argument is needed in the future anyway?

theoratkin committed:
Add interpolation parameter to resize_to_po2()

Image::resize_to_po2() now takes an optional p_interpolation parameter
that it passes directly to resize() with default value INTERPOLATE_BILINEAR.

theogen-ratkin commented:
Pushed the version that has `INTERPOLATE_BILINEAR` by default.

akien-mga commented on doc/classes/Image.xml:
```diff
@@ -430,8 +430,10 @@
+			</argument>
 			<description>
-				Resizes the image to the nearest power of 2 for the width and height. If [code]square[/code] is [code]true[/code] then set width and height to be the same.
+				Resizes the image to the nearest power of 2 for the width and height. If [code]square[/code] is [code]true[/code] then set width and height to be the same. New pixels are calculated using [code]interpolation[/code]. See [code]interpolation[/code] constants.
```
```suggestion
				Resizes the image to the nearest power of 2 for the width and height. If [code]square[/code] is [code]true[/code] then set width and height to be the same. New pixels are calculated using the [code]interpolation[/code] mode defined via [enum Interpolation] constants.
```

theogen-ratkin replied:
I just copied that bit from the `resize()` documentation above, so we should change it as well?

https://github.com/godotengine/godot/blob/053254466115eae12e48462c0c503280a21f1544/doc/classes/Image.xml#L415-L426

akien-mga replied:
Yeah that would be good. The part about `[code]interpolation[/code] constants` is incorrect, that's not a proper link to the enum.

theogen-ratkin replied:
Done, updated documentation of both methods.

akien-mga commented:
Thanks!
